pyro/contrib/gp/models/sgpr.py

```diff
@@ -80,64 +84,44 @@ def model(self):
-                trace_term += (Kffdiag - Qffdiag).sum() / self.noise
+                trace_term += (Kffdiag - Qffdiag).sum() / noise
 
         zero_loc = Variable(D.data.new([0])).expand(self.num_data)
         # DTC: cov = Qff + noise, trace_term = 0
         # FITC: cov = Qff + diag(Kff - Qff) + noise, trace_term = 0
         # VFE: cov = Qff + noise, trace_term = tr(Kff - Qff) / noise
-        pyro.sample("y", dist.SparseMultivariateNormal(zero_loc, D, W), obs=self.y)
+        pyro.sample("y", dist.SparseMultivariateNormal(zero_loc, D, W, trace_term), obs=self.y)
 
     def guide(self):
-        kernel_guide_prior = {}
-        for p in self.kernel_prior:
-            p_MAP_name = pyro.param_with_module_name(self.kernel.name, p) + "_MAP"
-            # init params by their prior means
-            p_MAP = pyro.param(p_MAP_name, Variable(self.kernel_prior[p].torch_dist.mean.data.clone(),
-                                                    requires_grad=True))
-            kernel_guide_prior[p] = dist.Delta(p_MAP)
-
-        kernel_fn = pyro.random_module(self.kernel.name, self.kernel, kernel_guide_prior)
-        kernel = kernel_fn()
-
-        Xu_guide_prior = {}
-        for p in self.Xu_prior:
-            p_MAP_name = pyro.param_with_module_name(self.Xu.name, p) + "_MAP"
-            # init params by their prior means
-            p_MAP = pyro.param(p_MAP_name, Variable(self.Xu_prior[p].torch_dist.mean.data.clone(),
-                                                    requires_grad=True))
-            Xu_guide_prior[p] = dist.Delta(p_MAP)
-
-        Xu_fn = pyro.random_module(self.Xu.name, self.Xu, Xu_guide_prior)
-        Xu = Xu_fn().inducing_points
-
-        return kernel, Xu
-
-    def forward(self, Xnew, full_cov=False, noiseless=False):
+        kernel = self.kernel.set_mode("guide")
+        self.set_mode("guide")
+        noise = self.get_param("noise")
+        Xu = self.get_param("Xu")
+
+        return kernel, noise, Xu
+
+    def forward(self, Xnew, full_cov=False, noiseless=True):
         """
-        Computes the parameters of `p(y*|Xnew) ~ N(loc, cov)`
-        w.r.t. the new input Xnew.
+        Computes the parameters of ``p(y*|Xnew) ~ N(loc, cov)`` w.r.t. the new input ``Xnew``.
 
         :param torch.autograd.Variable Xnew: A 2D tensor.
-        :param bool full_cov: Predict full covariance matrix or just its diagonal.
-        :param bool noiseless: Include noise in the prediction or not.
-        :return: loc and covariance matrix of p(y*|Xnew).
+        :param bool full_cov: Predicts full covariance matrix or just its diagonal.
+        :param bool noiseless: Includes noise in the prediction or not.
+        :return: loc and covariance matrix of ``p(y*|Xnew)``.
         :rtype: torch.autograd.Variable and torch.autograd.Variable
         """
         if Xnew.dim() == 2 and self.X.size(1) != Xnew.size(1):
             assert ValueError("Train data and test data should have the same feature sizes.")
```

Comment:
This should be `raise ValueError`; assert will never raise.

Reply:
Ah, my mistake, thanks!